db/view/view_update_generator.cc

```diff
@@ -24,6 +25,64 @@ static inline void inject_failure(std::string_view operation) {
 
 namespace db::view {
 
+class view_update_generator::progress_tracker final : public sstables::read_monitor_generator {
+    class read_monitor final : public  sstables::read_monitor {
+        sstables::shared_sstable _sst;
+        uint64_t& _pending_work;
+    public:
+        virtual void on_read_started(const sstables::reader_position_tracker& tracker) override {}
+
+        virtual void on_read_completed() override {
+            if (auto sst = std::exchange(_sst, {})) {
+                _pending_work -= sst->data_size();
+            }
+        }
+
+        read_monitor& operator=(const read_monitor&) = delete;
+        read_monitor(const read_monitor&) = delete;
+        read_monitor& operator=(const read_monitor&&) = delete;
+        read_monitor(read_monitor&&) = delete;
+
+        explicit read_monitor(sstables::shared_sstable sst, uint64_t& pw) : _sst(std::move(sst)), _pending_work(pw) {
+            _pending_work += _sst->data_size();
+        }
+        ~read_monitor() {
+            if (_sst) {
+                _pending_work -= _sst->data_size();
+            }
+        }
+    };
+private:
+    // read_monitor needs a stable ptr for operator().
```

Comment:
I didn't understand what this comment is suggesting. Is it trying to explain why we need shared_sstable? Why we need an "ordered_map"? What do you mean by a "stable" ptr or maybe this is just a typo for "sstable"?

Reply:
Actually stable. That's because operator() returns a reference to previously initialized monitor. If the ptr weren't stable, then we'd have a dangling ref.

Reply:
Will improve the comment.